### ISSUE TITLE
"304 Not Modified" is returning the wrong headers

### ISSUE THREAD

**Issue by michaellenaghan** [almarklein/asgineer#40]
There's a problem here:

https://github.com/almarklein/asgineer/blob/0b259a2806b3176592e699b93488f003f2d65aa3/asgineer/utils.py#L129

Take a look at this example:

```
$ curl -I -H "If-None-Match: 512dc2f89fea2c3794631fddb080e9db842542b0204cd3e2b13d1685d9f2a349" "http://localhost:5555/timetagger/"

HTTP/1.1 304 Not Modified
date: Wed, 28 Feb 2024 20:18:06 GMT
server: uvicorn
etag: 512dc2f89fea2c3794631fddb080e9db842542b0204cd3e2b13d1685d9f2a349
cache-control: public, must-revalidate, max-age=0
content-type: application/octet-stream
content-length: 0
```

Note the `content-type` (`application/octet-stream`) and `content-length` (`0`). 

Now take a look at the description of "304 Not Modified" on [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/304):

> The response must not contain a body and must include the headers that would have been sent in an equivalent [200](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/200) OK response: [Cache-Control](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control), [Content-Location](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Location), [Date](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Date), [ETag](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag), [Expires](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Expires), and [Vary](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Vary).

This error is causing a problem when a server is proxied behind Apache. I can explain the problem, but it's complicated and I'm not sure getting into the details will help. The short version, more or less, is that Apache appears to notice the discrepancy, and substitutes a 404 response.

**Comment by michaellenaghan** [comment]
Looking further at the code, the fix is very simple. Basically, just move the test down a little, after some of the headers have been filled in. I can submit a PR if you like, but there are some stylistic questions; it would probably be quicker for you to do it than for me to ask what you'd like!

**Comment by michaellenaghan** [comment]
> Looking further at the code, the fix is very simple.

OK, I take that back. :-)

I implemented a "very simple fix," and ran into this:

```
[ERROR 2024-02-28 21:35:12] LocalProtocolError in sending response: Too much data for declared Content-Length
Traceback (most recent call last):
  File "/home/protected/timetagger/.venv/lib/python3.9/site-packages/asgineer/_app.py", line 180, in _handle_http
    await request.send(body, more=False)
  File "/home/protected/timetagger/.venv/lib/python3.9/site-packages/asgineer/_request.py", line 213, in send
    await self._send(message)
  File "/home/protected/timetagger/.venv/lib/python3.9/site-packages/uvicorn/protocols/http/h11_impl.py", line 509, in send
    output = self.conn.send(event=h11.Data(data=data))
  File "/home/protected/timetagger/.venv/lib/python3.9/site-packages/h11/_connection.py", line 512, in send
    data_list = self.send_with_data_passthrough(event)
  File "/home/protected/timetagger/.venv/lib/python3.9/site-packages/h11/_connection.py", line 545, in send_with_data_passthrough
    writer(event, data_list.append)
  File "/home/protected/timetagger/.venv/lib/python3.9/site-packages/h11/_writers.py", line 65, in __call__
    self.send_data(event.data, write)
  File "/home/protected/timetagger/.venv/lib/python3.9/site-packages/h11/_writers.py", line 91, in send_data
    raise LocalProtocolError("Too much data for declared Content-Length")
h11._util.LocalProtocolError: Too much data for declared Content-Length
ERROR:    ASGI callable returned without completing response.
```

The uvicorn docs [say](https://www.uvicorn.org/server-behavior/#head-requests):

> If a `Content-Length` header is included then Uvicorn will ensure that the content length of the response body matches the value in the header, and raise an error otherwise.

I think that's what's happening?

There's (indirect) discussion of the issue [here](https://github.com/encode/starlette/issues/1764).

There's no *uvicorn* resolution, just (indirect) discussion.

I have, for the time being, commented out the 304 response.

**Comment by michaellenaghan** [comment]
Sorry about all of the spam...

I was able to avoid the error.

After moving the `etag` check down, I had factored out the setting of the `content-length`; I *always* set it, after `body` was set. And apparently that was wrong.

But...

But there's still a problem.

If I comment out the `etag` code, Apache is happy. 

If I leave it *in*, Apache is unhappy. 

I haven't figured out why. 

On my dev box, with direct access to Timetagger, the responses look fine.

In production, with Timetagger sitting behind Apache, I can't easily see the conversation between them, and sometimes Apache returns a 404.

I'll let you know if I find something useful. 

But it won't be tonight. :-)

**Comment by michaellenaghan** [comment]
Good heavens. More spam.

Apache is correct; the content-lengths are different:

```
< HTTP/1.1 200 OK
< date: Wed, 28 Feb 2024 22:35:17 GMT
< server: uvicorn
< etag: 512dc2f89fea2c3794631fddb080e9db842542b0204cd3e2b13d1685d9f2a349
< cache-control: public, must-revalidate, max-age=0
< content-type: text/html
< content-length: 9196
```

```
< HTTP/1.1 304 Not Modified
< date: Wed, 28 Feb 2024 22:36:02 GMT
< server: uvicorn
< etag: 512dc2f89fea2c3794631fddb080e9db842542b0204cd3e2b13d1685d9f2a349
< cache-control: public, must-revalidate, max-age=0
< content-type: text/html
< content-length: 9192
```

They're wrong — i.e., different — for HEAD requests too:

```
< HTTP/1.1 200 OK
< date: Wed, 28 Feb 2024 22:38:59 GMT
< server: uvicorn
< etag: 512dc2f89fea2c3794631fddb080e9db842542b0204cd3e2b13d1685d9f2a349
< cache-control: public, must-revalidate, max-age=0
< content-type: text/html
< content-length: 9192
```

**Comment by michaellenaghan** [comment]
OK, I'll take a look next week.

Btw, I realized that the error I saw earlier was directly related to this last finding, re: mis-matched `content-length`s. 

The error was "Too much data for declared Content-Length." `asgineer` says the content length is 9,192, while `h11` says the content length *of the same body* is 9,196. So, yes, when 9,192 is specified, there's "too much data." For some unknown reason.

Anyway, yes, OK, I'll take a look next week.

P.S. "Looking further at the code, the fix is very simple." Ha ha. Ha ha ha. You'd think I'd be old enough to know better by now... :-)

**Comment by almarklein** [comment]
>  asgineer says the content length is 9,192, while h11 says the content length of the same body is 9,196.

mmm. Sound like a bug in asgineer. Could this be a difference between the string length and the length in (utf-8 encoded) bytes? 

> P.S. "Looking further at the code, the fix is very simple." Ha ha. Ha ha ha. You'd think I'd be old enough to know better by now... :-)

The code to add is probably simple. Now it's just a matter of figuring out exactly *what* code 😉

**Comment by michaellenaghan** [comment]
Fixed in #41
